### ISSUE TITLE
Adding `Drop` impl to a type used in a public API type is a SemVer hazard

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#930]
If a type in the public API has no components that `impl Drop`, adding the first `impl Drop` on that type or any of its components appears to be a major breaking change: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5c132b6f967c798b59e2d5dbad533fe1)
```rust
struct Foo;

// Adding this block is a breaking change downstream on `Bar`.
impl Drop for Foo {
    fn drop(&mut self) {}
}

pub struct Bar(Foo);

// Witness of the breakage.
// We don't need to use the value, we're just dropping it.
const fn example(_value: Bar) -> i64 {
    0
}
```

I don't believe we have a reliable way of detecting this via rustdoc JSON right now. We can check for `impl Drop` on the public type itself, but I don't see a way to reliably check that for its components.

Perhaps either rustdoc JSON could be extended with a new field with this information, or a new marker auto-trait could be added to the language that we could track?

**Comment by djc** [comment]
> I don't believe we have a reliable way of detecting this via rustdoc JSON right now. We can check for `impl Drop` on the public type itself, but I don't see a way to reliably check that for its components.

To the extent that implementation of this lint has been held back by this, I would strongly suggest to avoid letting the perfect be the enemy of the good.

**Comment by obi1kenobi** [comment]
> avoid letting the perfect be the enemy of the good

Whenever we can avoid false-positives and still catch even just _some_ true-positives, this is indeed our guiding principle.

In this case however, I don't think that's the case. Consider the following example:
```rust
struct Foo;

impl Drop for Foo {
    fn drop(&mut self) {}
}

pub struct Bar(Foo);

// Adding this block is not breaking,
// because `Bar` already had drop glue
// because of `impl Drop for Foo`.
impl Drop for Bar {
    fn drop(&mut self) {}
}
```
If `Foo` came from a different crate, we can't know about its `Drop` impl, so we'll throw a (bad & very common) false-positive here when `Bar` gets a `Drop` impl too.

This recursive, cross-crate analysis (which is analogous to how auto-traits work, actually) is why we need to know about drop glue from rustdoc JSON (and rustc) and can't do it ourselves.

**Comment by willglynn** [comment]
I was reading https://github.com/rust-lang/rfcs/pull/3762 and found some context to add here.

Const traits will require reasoning about whether `Drop` is `const`. That body of work introduces `std::marker::Destruct` (https://github.com/rust-lang/rust/issues/133214). This is intended to allow users to add trait bounds on whether `Drop` is `const`, but may additionally be responsive to the broader semver breakage caused by the addition of drop glue.

The current RFC describes:

> The `Destruct` trait enables dropping types within a const context.
> 
> ```rust
> const fn foo<T>(t: T) {
>     // `t` is dropped here, but we don't know if we can evaluate its `Drop` impl (or that of its fields' types)
> }
> const fn baz<T: Copy>(t: T) {
>     // Fine, `Copy` implies that no `Drop` impl exists
> }
> const fn bar<T: [const] Destruct>(t: T) {
>     // Fine, we can safely invoke the destructor of `T`.
> }
> ```
> 
> When a value of a generic type goes out of scope, it is dropped and (if it has one) its `Drop` impl gets invoked. This situation seems no different from other trait bounds, except that types can be dropped without implementing `Drop` (as they can contain types that implement `Drop`). In that case the type's drop glue is invoked.
> 
> The `Destruct` trait is a bound for whether a type has drop glue. This is trivally true for all types.
> 
> `[const] Destruct` trait bounds are satisfied only if the type's `Drop` impl (if any) is `const` and all of the types of its components are `[const] Destruct`.
> 
> While this means that it's a breaking change to add a type with a non-const `Drop` impl to a type, that's already true and nothing new:
> 
> ```rust
> pub struct S {
>     x: u8,
>     y: Box<()>, // adding this field breaks code.
> }
> 
> const fn f(_: S) {}
> //~^ ERROR destructor of `S` cannot be evaluated at compile-time
> ```

**Comment by obi1kenobi** [comment]
Interesting, thank you for finding and summarizing the RFC!

You mentioned `Destruct` might be useful for detecting the addition of drop glue. Is there such a thing as "const drop glue"? Is the idea that the SemVer breakage caused by the addition of drop glue is exactly identical to "no longer implements `const Destruct`"?

**Comment by willglynn** [comment]
@obi1kenobi Yep! The motivation for that marker trait is to distinguish those types which can be dropped in a const context from those which can't.

The need for a separate trait wasn't immediately obvious to me. `std::marker::Destruct` is distinct from  `std::ops::Drop` because `Drop` refers to _custom_ drop code, rather than dropping in general.

The bound `T: Drop` means "`T` provides a custom `Drop` implementation". `T: [const] Drop` would therefore mean "`T` provides a custom `Drop` implementation which is callable in const contexts". What we need for the semver breakage or the kinds of bounds required for generics in a world with const traits is "`T` can be dropped in `const` contexts", irrespective of whether there is a custom const-callable `Drop` implementation.

It's not practical to change what `T: Drop` means, so instead, add `std::marker::Destruct`. `T: Destruct` means "`T` can be dropped", which is true for all types `T` and is therefore not a useful thing to say. However, by introducing `Destruct`, we've opened the door to `T: [const] Destruct`, which means "`T` can be dropped in const contexts", which _is_ the semver breakage motivating this issue.

**Comment by obi1kenobi** [comment]
Awesome, thank you!
